### ISSUE TITLE
Error while compiling with qt 5.8

### ISSUE THREAD

**Issue by simon-r** [OpenBoard-org/OpenBoard#78]
During the compilation of OpenBoard 1.3.4 whit qt 5.8 and g++ 6.3.1 I've had this error.

```
rc/core/UBTextTools.cpp: In static member function ‘static QString UBTextTools::cleanHtmlCData(const QString&)’:
src/core/UBTextTools.cpp:37:24: error: ambiguous overload for ‘operator!=’ (operand types are ‘const QChar’ and ‘char’)
         if(_html.at(i) != '\0')
            ~~~~~~~~~~~~^~~~~~~
In file included from /usr/include/qt/QtCore/qregexp.h:47:0,
                 from /usr/include/qt/QtCore/qstringlist.h:47,
                 from /usr/include/qt/QtCore/qlist.h:1093,
                 from /usr/include/qt/QtCore/qhash.h:46,
                 from /usr/include/qt/QtCore/qdebug.h:45,
                 from /usr/include/qt/QtCore/QDebug:1,
                 from src/core/UBTextTools.cpp:28:
/usr/include/qt/QtCore/qstring.h:1639:13: note: candidate: bool operator!=(QChar, const QStringRef&)
 inline bool operator!=(QChar lhs, const QStringRef &rhs) Q_DECL_NOTHROW { return !(lhs == rhs); }
             ^~~~~~~~
/usr/include/qt/QtCore/qstring.h:1625:13: note: candidate: bool operator!=(const QString&, QChar)
 inline bool operator!=(const QString &lhs, QChar rhs) Q_DECL_NOTHROW { return !(rhs == lhs); }
             ^~~~~~~~
/usr/include/qt/QtCore/qstring.h:1620:13: note: candidate: bool operator!=(QChar, const QString&)
 inline bool operator!=(QChar lhs, const QString &rhs) Q_DECL_NOTHROW { return !(lhs == rhs); }
             ^~~~~~~~
/usr/include/qt/QtCore/qstring.h:1567:13: note: candidate: bool operator!=(const QString&, const QStringRef&)
 inline bool operator!=(const QString &lhs, const QStringRef &rhs) Q_DECL_NOTHROW { return lhs.compare(rhs) != 0; }
             ^~~~~~~~
In file included from /usr/include/qt/QtCore/qhash.h:44:0,
                 from /usr/include/qt/QtCore/qdebug.h:45,
                 from /usr/include/qt/QtCore/QDebug:1,
                 from src/core/UBTextTools.cpp:28:
/usr/include/qt/QtCore/qchar.h:576:30: note: candidate: constexpr bool operator!=(QChar, QChar)
 Q_DECL_CONSTEXPR inline bool operator!=(QChar c1, QChar c2) Q_DECL_NOTHROW { return !operator==(c1, c2); }
                              ^~~~~~~~
g++ -c -pipe -fopenmp -O2 -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -w -D_REENTRANT -fPIC -DQUAZIP_STATIC -DNO_THIRD_PARTY_WARNINGS -DUBVERSION="\"1.3.4."\" -DUBVERSION_RC=1,3,4,240,0 -DQT_NO_DEBUG -DQT_WEBKITWIDGETS_LIB -DQT_WEBKIT_LIB -DQT_SVG_LIB -DQT_UITOOLS_LIB -DQT_MULTIMEDIAWIDGETS_LIB -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_MULTIMEDIA_LIB -DQT_GUI_LIB -DQT_XMLPATTERNS_LIB -DQT_NETWORK_LIB -DQT_XML_LIB -DQT_SCRIPT_LIB -DQT_CORE_LIB -D_REENTRANT -Isrc -I../OpenBoard-ThirdParty/zlib/1.2.8/include -I../OpenBoard-ThirdParty/xpdf/xpdf-3.04 -I../OpenBoard-ThirdParty/xpdf/xpdf-3.04/goo -I../OpenBoard-ThirdParty/xpdf/xpdf-3.04/splash -Isrc/web/browser -Isrc/pdf-merger -I../OpenBoard-ThirdParty/quazip -I../OpenBoard-ThirdParty/zlib/1.2.8/include -I../OpenBoard-ThirdParty/quazip/quazip-0.7.1 -I../OpenBoard-ThirdParty/qt/singleapplication -I../OpenBoard-ThirdParty/qt/singleapplication -I../OpenBoard-ThirdParty/qt/lockedfile -isystem /usr/include/qt -isystem /usr/include/qt/QtWebKitWidgets -isystem /usr/include/qt/QtWebKit -isystem /usr/include/qt/QtSvg -isystem /usr/include/qt/QtUiTools -isystem /usr/include/qt/QtMultimediaWidgets -isystem /usr/include/qt/QtPrintSupport -isystem /usr/include/qt/QtWidgets -isystem /usr/include/qt/QtMultimedia -isystem /usr/include/qt/QtGui -isystem /usr/include/qt/QtXmlPatterns -isystem /usr/include/qt/QtNetwork -isystem /usr/include/qt/QtXml -isystem /usr/include/qt/QtScript -isystem /usr/include/qt/QtCore -Ibuild/linux/release/moc -isystem /usr/include/libdrm -Ibuild/linux/release/ui -isystem /usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/lib/qt/mkspecs/linux-g++ -o build/linux/release/objects/UBDocumentController.o src/document/UBDocumentController.cpp
make[1]: *** [Makefile.Release:4200: build/linux/release/objects/UBTextTools.o] Error 1
make[1]: *** Attesa per i processi non terminati....
make[1]: uscita dalla directory "/tmp/yaourt-tmp-simone/aur-openboard/src/OpenBoard-1.3.4"
make: *** [Makefile:38: release] Error 2

```

**Comment by dynobo** [comment]
I got exactly same error with same versions of gcc, openboard and qt. (running arch)

**Comment by freeheel3r** [comment]
Thanks for reporting this. 

I don't have a Qt 5.8 / g++ 6 setup to test this on right now, but maybe this can be fixed by changing that line to the following:

        if(_html.at(i) != QChar('\0'))

Any chance one of you could test this?

P.S I edited your comment for readability

**Comment by dynobo** [comment]
I think this fixed the first problem. Now I ran into another error:
`
usr/lib/glib-2.0/include -I/usr/lib/qt/mkspecs/linux-g++ -o build/linux/release/objects/UBGraphicsMediaItem.o src/domain/UBGraphicsMediaItem.cpp
In file included from /usr/include/qt/QtMultimedia/QtMultimedia:77:0,
                 from src/domain/UBGraphicsMediaItemDelegate.h:35,
                 from src/domain/UBGraphicsMediaItem.cpp:32:
/usr/include/qt/QtMultimedia/qtmultimediadefs.h:1:51: schwerwiegender Fehler: ../../src/multimedia/qtmultimediadefs.h: Datei oder Verzeichnis nicht gefunden
 #include "../../src/multimedia/qtmultimediadefs.h"
                                                   ^
Kompilierung beendet.
make[1]: *** [Makefile.Release:4569: build/linux/release/objects/UBGraphicsMediaItem.o] Fehler 1
make[1]: Verzeichnis „/home/holger/Downloads/src/OpenBoard-master“ wird verlassen
make: *** [Makefile:38: release] Fehler 2
==> FEHLER: Ein Fehler geschah in build().
`
Seems to be an [qt issue](https://bugreports.qt.io/browse/QTBUG-58432)?

**Comment by stephensrmmartin** [comment]
The second issue seems to be due to Qt 5.8 deprecating qtmultimediadefs.h

But I'm not sure what one is supposed to do here? If you include <QtMultimedia>, it pulls in QtMultimedia.h, which in turn includes qtmultimediadefs.h, which #includes a non-existing file. 
I assume they want people to stop using QtMultimedia as an include, but rather include each separate module within QtMultimedia?

**Comment by freeheel3r** [comment]
Actually it seems like this issue was resolved by Qt (according to [this](https://bugreports.qt.io/browse/QTBUG-58432)).
It would be included in Qt 5.8.1, but apparently they might not offer a 5.8.1 version.. so unless you're building Qt from the development branches it may take a while for the fix to arrive.

**Comment by stephensrmmartin** [comment]
I assume it would be safe, in the interim, to temporarily delete the include in QtMultimedia that references that line then?

**Comment by freeheel3r** [comment]
Well.. the only advice I can offer is "there's one way to find out" :) 
I haven't actually tested all this myself, I'm still on Qt 5.5 for now.

**Comment by stephensrmmartin** [comment]
I removed the include "qtmultimediadefs.h" line from the QtMultimedia header and it worked fine. I think this is safe to do, since the qt team acknowledges that this should be gone anyway, and all other packages on my system seemed to be compiled just fine despite that issue.

So the patches referenced in #89 + removing that line in the qt 5.8 QtMultimedia.h header allowed openboard to compile just fine. I also think this will be mostly cross-compatible with other distros, because all I really did was modify the code as you mentioned here (QChar) and added an include path for openssl-1.0. The only non-compatible change was specifying that qmake should -l libcrypto.so.1.0.0 specifically, rather than just -lcrypto, which would by default link libcrypto.so, which is openssl 1.1's version. I did this as a quick fix, but I assume qmake has some way of specifying which library version to link against, in which case one can simply specify that the project needs openssl 1.0's version, not 1.1. I'm not knowledgeable at all about qmake, so I leave that to those who actually know what they're doing :). Thanks for the quick responses!

**Comment by stephensrmmartin** [comment]
The AUR maintainer for this package used this patch with success: https://aur.archlinux.org/cgit/aur.git/tree/qtmultimediadefs.patch?h=openboard

**Comment by fsiegert** [comment]
Just to mention, that the first part of the issue (QChar) should be fixed by pull request #132 . The second part of the issue (QtMultimedia includes) is probably obsolete by now as this has been fixed by Qt upstream in newer versions (apparently).
